### PR TITLE
Train package disease model

### PR DESCRIPTION
**Pull request by nsorros**
This PR adds the necessary code to train the baseline disease model based on Mesh. To do so

- it introduces a config file to run all steps for training the disease mesh model
- adds a config and the ability to parse params from a config in all related files
- adds y_batch_size param to train to split training on multiple labels to make it more memory friendly.

It lacks
* any changes that will be needed to package the model / code, pending discussion with @jdu as to how to package / ship the 75GB model

### CODE REVIEW

File: grants_tagger/train.py [hunk 1]
```diff
@@ -223,14 +224,44 @@ def train_and_evaluate(
...
         )
-        model.fit(X_train, Y_train)
+        if Y_train.shape[1] > 512:
+            vectorizer = model.steps[0][1]
+            classifier = model.steps[1][1]
+
+            Path(model_path).mkdir(exist_ok=True)
+            print("Fitting vectorizer")
+            vectorizer.fit(X_train)
+            with open(f"{model_path}/tfidf.pkl", "wb") as f:
```
Comment (nsorros) [inline grants_tagger/train.py:234]: this needs to be renamed vectorizer to generalise to keras_vectorizer for example

File: grants_tagger/train.py [hunk 2]
```diff
@@ -223,14 +224,44 @@ def train_and_evaluate(
             label_binarizer=label_binarizer,
             from_same_distribution=from_same_distribution
         )
-        model.fit(X_train, Y_train)
+        if Y_train.shape[1] > 512:
```
Comment (nsorros) [inline grants_tagger/train.py:227]: this flag possibly needs a different name since neural nets do not need batching of Y. possibly y_batch_size.

**Comment by nsorros** [comment]
[Deploy tfidf-svm baseline for disease Mesh](https://trello.com/c/b5b0rSwQ/1021-deploy-tfidf-svm-baseline-for-disease-mesh)

File: configs/disease_mesh/2020.07.0.ini [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+[data]
+train_data_path=data/processed/disease_mesh.jsonl
+test_data_path=
+
+[model]
+label_binarizer_path=models/disease_mesh_label_binarizer.pkl
+model_path=models/disease_mesh_tfidf-svm-2020.07.0
+approach=tfidf-sgd
+parameters={"tfidf__ngram_range": (1,2), "svm__estimator__loss": "log", "svm__estimator__alpha": 1e-9, "tfidf__max_features": 2_000_000}
```
Comment (aCampello) [inline configs/disease_mesh/2020.07.0.ini:21]: why underscore (`2_000_000`) ?

File: grants_tagger/filter_disease_codes.py [hunk 4]
```diff
@@ -29,6 +30,20 @@ def filter_disease_codes(mesh_descriptions_file, mesh_export_file):
...
+    argparser.add_argument('--config', type=Path, help="path to config file defining arguments")
     args = argparser.parse_args()
 
-    filter_disease_codes(args.mesh_descriptions_file, args.mesh_export_file)
+    if args.config:
+        config_parser = ConfigParser()
+        cfg = config_parser.read(args.config)
+
+        mesh_descriptions_file = cfg["filter_disease_codes"]["mesh_descriptions_file"]
+        mesh_export_file = cfg["filter_disease_codes"][""]
```
Comment (aCampello) [inline grants_tagger/filter_disease_codes.py:41]: Is this correct? (the last bracket `[""]`)

File: grants_tagger/preprocess_mesh.py [hunk 5]
```diff
@@ -1,13 +1,20 @@
...
 import json
+import os
+
+import pandas as pd
 
 def yield_raw_data(input_path):
     with open(input_path, encoding='latin-1') as f_i:
-        for line in f_i:
+        for i, line in enumerate(f_i):
+            if i == 0:
```
Comment (aCampello) [inline grants_tagger/preprocess_mesh.py:15]: Perhaps `f_i.readline()` before the loop is cleaner?

File: grants_tagger/train.py [hunk 6]
```diff
@@ -223,14 +225,44 @@ def train_and_evaluate(
...
+            Path(model_path).mkdir(exist_ok=True)
+            print("Fitting vectorizer")
+            vectorizer.fit(X_train)
+            with open(f"{model_path}/vectorizer.pkl", "wb") as f:
+                f.write(pickle.dumps(vectorizer))
+            print("Training model")
+            for tag_i in range(0, Y_train.shape[1], y_batch_size):
+                print(tag_i)
+                X_train_vec = vectorizer.transform(X_train)
+                classifier.fit(X_train_vec, Y_train[:,tag_i:tag_i+512]) # assuming that fit clears previous params
```
Comment (aCampello) [inline grants_tagger/train.py:241]: should this be `Y_train[:,tag_i:tag_i+y_batch_size]` ? I am not sure if I understand the 512.

File: grants_tagger/train.py [hunk 7]
```diff
@@ -223,14 +225,44 @@ def train_and_evaluate(
...
+            pass # to be implemented
+        else:
+            Y_pred_prob = model.predict_proba(X_test)
         Y_pred_test = Y_pred_prob > threshold
     else:
-        Y_pred_test = model.predict(X_test)
-    #    Y_pred_train = model.predict(X_train)
+        if Y_test.shape[1] > 512:
+            Y_pred_test = []
+            for tag_i in range(0, Y_test.shape[1], 512):
```
Comment (aCampello) [inline grants_tagger/train.py:256]: Same as before - where did 512 come from? Is it a default batch size?

File: scripts/run_mesh_config.sh [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+CONFIG="${SCRIPT_DIR}/../configs/disease_mesh/${VERSION}.ini"
+
+# preprocess
+venv/bin/python -m grants_tagger.preprocess_mesh --config ${CONFIG}
+
+# create label binarizer
+venv/bin/python -m grants_tagger.create_mesh_label_binarizer --config ${CONFIG}
+
+# train
+venv/bin/python -m grants_tagger.train --config ${CONFIG}
```
Comment (aCampello) [inline scripts/run_mesh_config.sh:17]: Nice file to run the pipeline

File: grants_tagger/train.py (hunk 6 above)
Comment (nsorros) [reply grants_tagger/train.py:241]: i will replace it with `y_batch_size`, is that clearer?

File: grants_tagger/train.py (hunk 7 above)
Comment (nsorros) [reply grants_tagger/train.py:256]: same treatment as above

File: grants_tagger/preprocess_mesh.py (hunk 5 above)
Comment (nsorros) [reply grants_tagger/preprocess_mesh.py:15]: good idea

File: grants_tagger/filter_disease_codes.py (hunk 4 above)
Comment (nsorros) [reply grants_tagger/filter_disease_codes.py:41]: Hmm, that seems weird, let me investigate

File: configs/disease_mesh/2020.07.0.ini (hunk 3 above)
Comment (nsorros) [reply configs/disease_mesh/2020.07.0.ini:21]: why 2M or why I am using the underscore?

File: grants_tagger/preprocess_mesh.py (hunk 5 above)
Comment (nsorros) [reply grants_tagger/preprocess_mesh.py]: Resolved by 95920652fba6f9458ca653d401501a62404bcc99

File: grants_tagger/train.py (hunk 7 above)
Comment (nsorros) [reply grants_tagger/train.py]: Resolved by 2701cf6ab0b2232256378fe5a327d940f242c36e

File: scripts/run_mesh_config.sh (hunk 8 above)
Comment (nsorros) [reply scripts/run_mesh_config.sh:17]: 👍

File: grants_tagger/train.py (hunk 6 above)
Comment (nsorros) [reply grants_tagger/train.py]: Resolved by 2701cf6ab0b2232256378fe5a327d940f242c36e

File: configs/disease_mesh/2020.07.0.ini (hunk 3 above)
Comment (nsorros) [reply configs/disease_mesh/2020.07.0.ini:21]: underscore is ignored by python, it is just for readability

File: grants_tagger/filter_disease_codes.py (hunk 4 above)
Comment (nsorros) [reply grants_tagger/filter_disease_codes.py]: Good find, not on purpose.

Resolved by 45ebc17c0cc4fb413c283dd0f6a664bf4f0d2368

File: configs/disease_mesh/2020.07.0.ini (hunk 3 above)
Comment (aCampello) [reply configs/disease_mesh/2020.07.0.ini:25]: ha - didn't know that

**Comment by nsorros** [comment]
Addressed Antonio's comments. This is pending only a conversation with @jdu about how to effectively package the 75GB model which might not require code at all. Will address any code changes needed separately.
